cirq-core/cirq/contrib/zxtransformer/zxtransformer.py

```diff
@@ -0,0 +1,193 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A custom transformer for Cirq which uses ZX-Calculus for circuit optimization, implemented using PyZX."""
+
+from typing import Dict, List, Callable, Optional, Union
+
+import cirq
+from cirq import circuits
+
+import pyzx as zx
+from pyzx.circuit import gates as zx_gates
+
+
+def cirq_gate_to_zx_gate(cirq_gate: Optional[cirq.Gate], qubits: List[int]) -> Optional[zx_gates.Gate]:
+    """Convert a Cirq gate to a PyZX gate."""
+    if cirq.Gate is None:
+        return None
+
+    if isinstance(cirq_gate, (cirq.Rx, cirq.XPowGate)):
+        return zx_gates.XPhase(*qubits, phase=cirq_gate.exponent)  # type: ignore
+    if isinstance(cirq_gate, (cirq.Ry, cirq.YPowGate)):
+        return zx_gates.YPhase(*qubits, phase=cirq_gate.exponent)  # type: ignore
+    if isinstance(cirq_gate, (cirq.Rz, cirq.ZPowGate)):
+        return zx_gates.ZPhase(*qubits, phase=cirq_gate.exponent)  # type: ignore
+
+    # TODO: Deal with exponents other than nice ones.
+    if isinstance(cirq_gate, cirq.HPowGate):
+        if cirq_gate.exponent != 1.0:
+            raise ValueError(f"Unsupported exponent: {cirq_gate}.")
+        return zx_gates.HAD(*qubits)
+    if isinstance(cirq_gate, cirq.CZPowGate):
+        if cirq_gate.exponent != 1.0:
+            raise ValueError(f"Unsupported exponent: {cirq_gate}.")
+        return zx_gates.CZ(*qubits)
+    if isinstance(cirq_gate, cirq.CNotPowGate):  # maybe CXPowGate?
+        if cirq_gate.exponent != 1.0:
+            raise ValueError(f"Unsupported exponent: {cirq_gate}.")
+        return zx_gates.CNOT(*qubits)
+    if isinstance(cirq_gate, cirq.SwapPowGate):
+        if cirq_gate.exponent != 1.0:
+            raise ValueError(f"Unsupported exponent: {cirq_gate}.")
+        return zx_gates.SWAP(*qubits)
+    if isinstance(cirq_gate, cirq.CCZPowGate):
+        if cirq_gate.exponent != 1.0:
+            raise ValueError(f"Unsupported exponent: {cirq_gate}.")
+        return zx_gates.CCZ(*qubits)
+
+    return None
+
+
+cirq_gate_table = {
+    'rx':   cirq.XPowGate,
+    'ry':   cirq.YPowGate,
+    'rz':   cirq.ZPowGate,
+    'h':    cirq.HPowGate,
+    'cx':   cirq.CXPowGate,
+    'cz':   cirq.CZPowGate,
+    'swap': cirq.SwapPowGate,
+    'ccz':  cirq.CCZPowGate,
+}
+
+
+def _optimize(c: zx.Circuit) -> zx.Circuit:
+    g = c.to_graph()
+    zx.simplify.full_reduce(g)
+    return zx.extract.extract_circuit(g)
+
+
+@cirq.transformer
+class ZXTransformer:  # pylint: disable=too-few-public-methods
+    """Transformer to do processing on an input circuit with pyzx."""
+
+    def __init__(self, optimize: Optional[Callable[[zx.Circuit], zx.Circuit]] = None):
+        """Initializes transformer.
+
+        Args:
+            optimize: The optimization routine to execute. Defaults to `pyzx.simplify.full_reduce` if not specified.
+        """
+
+        super().__init__()
+        self.qubits: List[cirq.Qid] = []
```

Comment:
I don't think the qubits and qubit_to_index map should be properities of the class. they are properties of the circuit being transformed so they shouldn't be stored this way. 

Actually I don't think this transfomer need to be class. I think it should be a function
```py3
@transformer
def zx_transfomer(circuit: circuits.AbstractCircuit, context: Optional[cirq.TransformerContext] = None, optimizer: Callable[[zx.Circuit], zx.Circuit] = _optimize) -> circuits.AbstractCircuit
```
The qubits and qubit_to_index lists should be passed as arguments to the methods that need them.